### PR TITLE
⏰submit timezone info when signing up

### LINKED ISSUE #3724: Detect timezone on signup
And set the timezone as the user's time zone in the preferences, so that reports look correctly from the start.

### PR DESCRIPTION
**Pull request by khoi**
## What's this?
Send user's timezone data to /signup endpoint 

### Relationships
Closes #3724

## Why do we want this?
🕰

## How is it done?
I checked the implementation on the web and mimic it. The steps are:

- Get supported timezone from `/timezones` endpoint
- Get the device time zone 
- Check if the device timezone is supported if yes send it up, else send blank and backend will set it to the default value "utc/etc"

### Side effects
I hope there is none.

## Review hints
None beside testing it.

## Tests
Integration tests: https://app.bitrise.io/build/20ad4fa857a7a9b8

## :squid: Permissions
🦑

### CODE REVIEW

File: Toggl.Foundation/Interactors/Timezones/GetSupportedTimezonesInteractor.cs [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+        public GetSupportedTimezonesInteractor(ITogglApi api)
+        {
+            Ensure.Argument.IsNotNull(api, nameof(api));
+
+            this.api = api;
+        }
+
+        public IObservable<List<string>> Execute()
+            => api.Timezones.GetAll();
```
Comment (simonrozsival) [inline Toggl.Foundation/Interactors/Timezones/GetSupportedTimezonesInteractor.cs]: I don't think we should use an API call for this. Let's do it the same way we do it for countries list please.

File: Toggl.Foundation.MvvmCross/ViewModels/EditDurationViewModel.cs [hunk 2]
```diff
@@ -278,6 +278,11 @@ public override void ViewDestroy(bool viewFinishing)
             runningTimeEntryDisposable?.Dispose();
         }
 
+        public override void ViewDisappeared()
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/ViewModels/EditDurationViewModel.cs]: ?

File: Toggl.Foundation.MvvmCross/ViewModels/SignupViewModel.cs [hunk 3]
```diff
@@ -217,6 +225,22 @@ public override async Task Initialize()
                         getCountrySubscription = null;
                     }
                 );
+
+            var timezoneInteractor = new GetSupportedTimezonesInteractor();
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/ViewModels/SignupViewModel.cs]: Please don't `new` inside our ViewModels, specially with interactors. We have factories for these

File: Toggl.Daneel/Services/TimezoneServiceIos.cs [hunk 4]
```diff
@@ -0,0 +1,12 @@
+using System;
+using System.Reactive.Linq;
+using Foundation;
+using Toggl.Foundation.MvvmCross.Services;
+
+namespace Toggl.Daneel.Services
+{
+    public sealed class TimezoneServiceIos : ITimezoneService
+    {
+        public IObservable<string> DeviceTimezone => Observable.Return(NSTimeZone.LocalTimeZone.Name);
```
Comment (heytherewill) [inline Toggl.Daneel/Services/TimezoneServiceIos.cs]: This operation is not async. Why do you need that much boilerplate for making it a service of its own? It'd be much better as a property of the existing `IPlatformInfo` interface.

File: Toggl.Foundation.MvvmCross/ViewModels/SignupViewModel.cs [hunk 5]
```diff
@@ -217,6 +225,22 @@ public override async Task Initialize()
...
+            var supportedTimezonesObs = timezoneInteractor.Execute();
+
+            supportedTimezonesObs
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/ViewModels/SignupViewModel.cs]: Do we really wanna fire this before and store it into a property instead of doing it all when user signs up?

File: Toggl.Foundation/Interactors/Timezones/GetSupportedTimezonesInteractor.cs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Toggl.Foundation.Interactors.Timezones
+{
+    public sealed class GetSupportedTimezonesInteractor : IInteractor<IObservable<List<string>>>
+    {
+        public IObservable<List<string>> Execute()
+        {
+            string json = Resources.TimezonesJson;
+
+            var timezones = new Serialization
```
Comment (heytherewill) [inline Toggl.Foundation/Interactors/Timezones/GetSupportedTimezonesInteractor.cs]: Please inject the serializer instead of instantiating it

File: Toggl.Ultrawave/TogglApi.cs [hunk 7]
```diff
@@ -51,6 +53,7 @@ public TogglApi(ApiConfiguration configuration, IApiClient apiClient)
         public IProjectsSummaryApi ProjectsSummary { get; }
         public IWorkspaceFeaturesApi WorkspaceFeatures { get; }
         public IFeedbackApi Feedback { get; }
+        public ITimeZonesApi Timezones { get; }
```
Comment (heytherewill) [inline Toggl.Ultrawave/TogglApi.cs]: Where are you using these?

File: Toggl.Foundation.MvvmCross/ViewModels/SignupViewModel.cs (hunk 3 above)
Comment (khoi) [reply Toggl.Foundation.MvvmCross/ViewModels/SignupViewModel.cs]: hmm, I was following the implementation for `GetAllCountriesInteractor`, is there reason that interactor isn't in the factory? Or should it also be?

File: Toggl.Ultrawave/TogglApi.cs (hunk 7 above)
Comment (khoi) [reply Toggl.Ultrawave/TogglApi.cs]: Initially, the supported tz list comes from the API, but now I've hardcoded the JSON response into our `Resources.resx`

However, I'm using it in an integration test which is helpful to make sure that our hardcoded JSON is not stale. So I think we should keep the API here.

File: Toggl.Daneel/Toggl.Daneel.csproj [hunk 8]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (heytherewill) [inline Toggl.Daneel/Toggl.Daneel.csproj]: Undo this change

File: Toggl.Foundation.MvvmCross/MvvmCrossFoundation.cs [hunk 9]
```diff
@@ -41,7 +41,6 @@ public struct MvvmCrossFoundation
         public IIntentDonationService IntentDonationService { get; }
         public IPrivateSharedStorageService PrivateSharedStorageService { get; }
         public IRxActionFactory RxActionFactory { get; }
-
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/MvvmCrossFoundation.cs]: Undo this change

File: Toggl.Foundation.MvvmCross/MvvmCrossFoundation.cs [hunk 10]
```diff
@@ -118,7 +117,6 @@ public class Builder
             public IPermissionsService PermissionsService { get; private set; }
             public ICalendarService CalendarService { get; private set; }
             public IRxActionFactory RxActionFactory { get; private set; }
-
```
Comment (heytherewill) [inline Toggl.Foundation.MvvmCross/MvvmCrossFoundation.cs]: Undo this change

File: Toggl.Foundation.Tests/MvvmCross/ViewModels/BaseViewModelTests.cs [hunk 11]
```diff
@@ -43,7 +43,6 @@ public abstract class BaseViewModelTests<TViewModel> : BaseMvvmCrossTests
 
         protected TestScheduler TestScheduler { get; }
         protected IRxActionFactory RxActionFactory { get; }
-
```
Comment (heytherewill) [inline Toggl.Foundation.Tests/MvvmCross/ViewModels/BaseViewModelTests.cs]: Undo this change

File: Toggl.Giskard/Toggl.Giskard.csproj [hunk 12]
```diff
@@ -1,4 +1,4 @@
-<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+﻿<Project DefaultTargets="Build" ToolsVersion="4.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
```
Comment (heytherewill) [inline Toggl.Giskard/Toggl.Giskard.csproj]: Undo this change

File: Toggl.Ultrawave.Tests.Integration/Toggl.Ultrawave.Tests.Integration.csproj [hunk 13]
```diff
@@ -1,4 +1,4 @@
-<Project Sdk="Microsoft.NET.Sdk">
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (heytherewill) [inline Toggl.Ultrawave.Tests.Integration/Toggl.Ultrawave.Tests.Integration.csproj]: Undo this change
